File: Data Structures & Algorithms/subsets/submission-4.java
class Solution {
    List<List<Integer>> res;
    List<Integer> subset;

    private void dfs(int[] nums, int i) {
        if (i == nums.length) {
            res.add(new ArrayList<Integer>(subset));
            return;
        }

        subset.add(nums[i]);
        dfs(nums, i + 1);
        subset.remove(subset.size() - 1);

        dfs(nums, i + 1);
    }

    public List<List<Integer>> subsets(int[] nums) {
        res = new ArrayList<>();
        subset = new ArrayList<>();

        dfs(nums, 0);

        return res;
    }
}

File: Data Structures & Algorithms/target-sum/submission-4.py
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        def dfs(i, currSum):
            if i == len(nums):
                return 1 if target == currSum else 0

            return (dfs(i + 1, currSum + nums[i]) +
                    dfs(i + 1, currSum - nums[i]))

        return dfs(0, 0)